dali/kernels/imgproc/convolution/laplacian_cpu.h

```diff
@@ -0,0 +1,299 @@
+    no_smoothing_ = IsNoSmooting(window_sizes);
+    if (no_smoothing_) {
+      if (!single_dim_impl_) {
+        multi_dim_impl_.reset();
+        single_dim_impl_ = std::make_unique<SingleDimConv>();
+      }
+      return single_dim_impl_->Setup(ctx, in_shape, window_sizes[deriv_axis]);
+    }
+    if (!multi_dim_impl_) {
+      single_dim_impl_.reset();
+      multi_dim_impl_ = std::make_unique<MultiDimConv>();
+    }
+    return multi_dim_impl_->Setup(ctx, in_shape, window_sizes);
+  }
+
+  void Run(KernelContext& ctx, const TensorView<StorageCPU, Out, ndim> &out,
+           const TensorView<StorageCPU, const In, ndim>& in,
+           const std::array<TensorView<StorageCPU, const W, 1>, axes>& windows,
+           const T& transform = {}) {
+    if (no_smoothing_) {
+      single_dim_impl_->Run(ctx, out, in, windows[deriv_axis], transform);
+    } else {
+      multi_dim_impl_->Run(ctx, out, in, windows, transform);
+    }
+  }
+
+  bool no_smoothing_;
+  std::unique_ptr<MultiDimConv> multi_dim_impl_;
+  std::unique_ptr<SingleDimConv> single_dim_impl_;
+};
+
+
+/**
+ * @brief Provides Laplacian specializations for 2 and 3 dimensional data.
+ *  Run methods expect additional ``acc`` buffer that will be used to accumulate
+ *  partial derivatives. If ``Intermediate`` and ``Out`` are the same type,
+ *  the `acc` and ``out`` can be the same tensor.
+ */
+template <typename T, typename Intermediate, typename Out, typename In, typename W, int axes,
+          bool has_channels>
+struct LaplacianCPUBase;
+
+template <typename T, typename Intermediate, typename Out, typename In, typename W,
+          bool has_channels>
+struct LaplacianCPUBase<T, Intermediate, Out, In, W, 2, has_channels> {
+  static constexpr int axes = 2;
+  using DxKernel = Convolution<Intermediate, In, W, axes, 0, has_channels,
+                               TransScaleSat<Intermediate, W>>;
+  using DyKernel = Convolution<Out, In, W, axes, 1, has_channels, T>;
+  static constexpr int ndim = DyKernel::ndim;  // Dx ndim is the same
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorShape<ndim>& in_shape,
+                           const std::array<std::array<int, axes>, axes>& window_sizes) {
+    KernelRequirements req;
+    req.output_shapes.push_back(uniform_list_shape<ndim>(1, in_shape));
+
+    auto req_dx = sobel_dx_.Setup(ctx, in_shape, window_sizes[0]);
+    auto req_dy = sobel_dy_.Setup(ctx, in_shape, window_sizes[1]);
+
+    // Calculate max scratch memory required by sub-kernels
+    sub_scratch_sizes_ = MaxScratchSize(req_dx.scratch_sizes, req_dy.scratch_sizes);
+    req.scratch_sizes = sub_scratch_sizes_;
+
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorView<StorageCPU, Out, ndim> &out,
+           const TensorView<StorageCPU, Intermediate, ndim> &acc,
+           const TensorView<StorageCPU, const In, ndim>& in,
+           const std::array<std::array<TensorView<StorageCPU, const W, 1>, axes>, axes>& windows,
```

Comment:
I'm thinking that we might need some docs about the nesting of windows at least :)

Reply:
I've added a brief docstr to the LaplacianCPU declaration that describes the ordering used in windows-related arguments.